auth/userModel.js

```diff
@@ -0,0 +1,40 @@
+const mongoose = require('mongoose');
+const bcrypt = require('bcrypt');
+
+const userSchema = new mongoose.Schema({
+  username: {
+    type: String,
+    unique: true,
+    required: true,
+    lowercase: true
+  },
+  password: {
+    type: String,
+    required: true,
+    minlength: 4
+  }
+});
+
+userSchema.pre('save', function(next) {
+  console.log('pre save hook');
+  bcrypt.hash(this.password, 16, (err, hash) => {
+    if (err) next(err);
+    else {
+      this.password = hash;
+      next();
+    };
+  });
+});
+
+//from https://www.mongodb.com/blog/post/password-authentication-with-mongoose-part-1
+userSchema.methods.comparePassword = function(candidatePassword, cb) {
+    bcrypt.compare(candidatePassword, this.password, function(err, isMatch) {
+        if (err) return cb(err);
+        cb(null, isMatch);
+    });
+};
```

Comment:
I see that this comparePassword code is from that blog post, but I have 2 questions.
1 - Why did you choose to write it this way? 
2 - Why do you feel this is better than the way we've been doing it?

Reply:
It's not necessarily that I feel it's *better*, but I do like callback functions.  And the logic of "if the password matches, then do the callback function" is very comfortable to me.  The other way you're referring to is just returning a true or false, right?  So this way basically just saves me the trouble of having to do an if (comparePassword(pass)) {dostuff} and lets me instead comparePassword(pass, dostuff).

Reply:
Aha. Makes sense, thank you :smile: